impress-org/givewp-next-gen
# Feature: add compatibility for PayPal Standard

Related: https://github.com/impress-org/givewp/pull/6684

## Description



This adds support for the PayPal standard gateway in our next gen forms.

This gateway has been implemented as a fully backwards compatible gateway while our next gen feature plugin is installed.  This means the PayPal standard gateway that is potentially already being used on legacy form - will still work there but will also now work on next gen forms.  There is nothing extra to enable or setup.

To achieve this, next gen is unregistering the core paypal gateway, and registering a new `PayPalStandardGateway` that extends the core one.  So all the same methods will work as expected on legacy, but next gen adds a few different conditionals for redirecting to the proper location and handling redirects.

## Affects



- The PayPal standard gateway, _only_ when next gen is installed.
- Introduces a new trait that can override what's in core `HandleHttpResponses` for our form's purposes. 
- Updates the form js to handle redirects properly

## Visuals



**Completing a payment**

![2023-01-27 11 20 22](https://user-images.githubusercontent.com/10138447/215137750-a78c7fe7-9a54-41a6-8a9b-240c2ba749bc.gif)

**Cancelling a payment**

*Note: our form does not have a sense of flash messages yet, but at least we can bring them back to the form.

![2023-01-27 11 15 09](https://user-images.githubusercontent.com/10138447/215137011-e397a27c-550e-4bfe-8492-ec3a7e82592b.gif)

## Testing Instructions



- Setup PayPal standard in GiveWP (you'll need a developer & sandbox account)
- Donate using PayPal standard on a next gen form
- Donate using PayPal standard on a legacy form (make sure it works as expected)

**Note when using legacy form**
At the time of writing, the order of our gateway registration does not account for unregistering a gateway.
This results in seeing the paypal fields twice on the legacy form.  There is a [PR ](https://github.com/impress-org/givewp/pull/6684) in GiveWP core to resolve that issue.

## Pre-review Checklist
